Simplify multiple-of-element size access to arrays

tautschnig opened the pull request:
Array operations may fall back to byte_extract or byte_update expressions in parts of the code base. Simplify this to index or WITH expressions, respectively, when the offset is known to be a multiple of the array-element size.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Simplify multiple-of-element size access to arrays

Array operations may fall back to byte_extract or byte_update
expressions in parts of the code base. Simplify this to index or WITH
expressions, respectively, when the offset is known to be a multiple of
the array-element size, or a constant offset plus a multiple of the
array-element size.

Fixes: #8617

remi-delmas-3000 commented on src/util/pointer_offset_size.cpp:
```diff
@@ -687,7 +687,58 @@ std::optional<exprt> get_subexpression_at_offset(
   const auto offset_bytes = numeric_cast<mp_integer>(offset);
 
   if(!offset_bytes.has_value())
```
does `has_value` mean "literal value" ?

tautschnig replied:
Comment added.

remi-delmas-3000 commented on src/util/pointer_offset_size.cpp:
```diff
@@ -687,7 +687,58 @@ std::optional<exprt> get_subexpression_at_offset(
+      if(
+        !target_size_bits.has_value() || !elem_size_bits.has_value() ||
+        *elem_size_bits <= 0 ||
+        *elem_size_bits % config.ansi_c.char_width != 0 ||
```
Does this still allow the optimization to apply on arrays of type T where alignof(T) > 8  ?

tautschnig replied:
Yes, this shouldn't restrict on alignment constraints, but I'll fix up the comment to make clear this isn't constrained to single-byte-alignment.

remi-delmas-3000 commented on src/util/pointer_offset_size.cpp:
```diff
@@ -687,7 +687,58 @@ std::optional<exprt> get_subexpression_at_offset(
+        return {};
+      }
+
+      if(
```
bail if expression is not of the form `K * i`

remi-delmas-3000 commented on src/util/simplify_expr.cpp:
```diff
@@ -1870,7 +1872,7 @@ simplify_exprt::simplify_byte_extract(const byte_extract_exprt &expr)
 
   if(
-    bits.has_value() &&
+    offset.has_value() && bits.has_value() &&
```
What is this doing exactly ? seems like it restricts the optimization to a subset of cases ?

tautschnig replied:
That's just because that very restriction previously was placed earlier in the same function, but now we have some cases that we want to handle later in the same function (where `offset` does not need to be a constant).

remi-delmas-3000 commented on src/util/simplify_expr.cpp:
```diff
@@ -1986,7 +1988,9 @@ simplify_exprt::simplify_byte_extract(const byte_extract_exprt &expr)
     const auto &element_bit_width =
       pointer_offset_bits(array_type.element_type(), ns);
-    if(element_bit_width.has_value() && *element_bit_width > 0)
+    if(
```
again it seems like the optimization would apply less often ?

tautschnig replied:
That's just because that very restriction previously was placed earlier in the same function, but now we have some cases that we want to handle later in the same function (where `offset` does not need to be a constant).

remi-delmas-3000 commented on src/util/simplify_expr.cpp:
```diff
@@ -2227,14 +2231,106 @@ simplify_exprt::simplify_byte_update(const byte_update_exprt &expr)
+        return changed(simplify_rec(result_expr));
+      }
+      else if(
+        offset.id() == ID_mult && offset.operands().size() == 2 &&
```
this optimizes the encoding of an array update, and the very similar code above optimized a read ?

tautschnig replied:
Similar, yes, but still different enough to make anything except for the pattern-matching piece rather cumbersome to share.

remi-delmas-3000 commented on src/util/pointer_offset_size.cpp:
```diff
@@ -687,7 +687,58 @@ std::optional<exprt> get_subexpression_at_offset(
 
   if(!offset_bytes.has_value())
-    return {};
+  {
```
Is this the array read encoding ?

tautschnig replied:
Yes.

remi-delmas-3000 commented:
Is there a way to share some of the pattern matching conditions between the array-read and array-write cases ?

remi-delmas-3000 commented:
Seems like we should also be able to pattern match an access to a member of a struct in an array of structs like a[i].m, the offset would be of the form i*sizeof(T) + offset(T,m) and turn that into a functional update as well ?

tautschnig commented:
> Seems like we should also be able to pattern match an access to a member of a struct in an array of structs like a[i].m, the offset would be of the form i*sizeof(T) + offset(T,m) and turn that into a functional update as well ?

Done.

tautschnig committed:
Tests to demonstrate need for offset pattern matching

Although the tests produce correct verification results, the encoding
can be very large as shown in #8617.

remi-delmas-3000 reviewed (commented):
From my understanding of the code the cases where we have a symbolic offset value with some shape that does not properly align with the underlying type of the accessed object seem to be handled, but could we add some negative tests for that ?

i.e. tests that check that the array access or array update still reduce to byte-level operations when they are supposed to ?

Also I just thought of the case you mentioned is frequent in the linux kernel, where people access struct fields using a `base_pointer` + `offsetof(T, field)`, or accessing some array from its end pointer with a negative offset, which I don't think are handled.
